Show styles from userstyles.org in Stylus popup

derv82 opened the pull request:
Feature request / enhancement issue: https://github.com/openstyles/stylus/issues/243

Notes on the first revision of the PR:

1. I installed `editorconfig` and made my changes using *VisualStudio Code*. Let me know if the formatting is off...
2. `searchResults.js` is encapsulated to only expose `UserStylesAPI` and `SearchResults`
3. Before trying to install the style (`searchResults.js:215`), I fetch the Style object from userstyles.org's API. Sample request: https://userstyles.org/api/v1/styles/70271
   * Should I pull the `style` from somewhere else on the site? E.g. the UserScripts Greasemonkey/Tampermonkey code?
   * Would it be easier to load the actual HTML style page in a hidden iframe, and install it from there?

derv82 commented on popup/searchResults.js:
```diff
@@ -0,0 +1,251 @@
+            reason: 'update',
+            notify: true
+          });
+          saveStyleSafe(styleObject);
```
Here's where I (incorrectly) try to install the style. Comments explain my reasoning: I'm trying to pass a StyleObject from userstyles.org's API (it has [a specific format](https://userstyles.org/api/v1/styles/70271)) into the `saveStyleSafe` method, which expects some other format for input.

tophf replied:
Direct saving requires sectioned JSON which we fetch from links like `https://userstyles.org/styles/chrome/37035.json` or `https://userstyles.org/styles/chrome/37035.json?ik-syntax-theme=ik-Twilight&ik-syntax-codemirror=ik-cm-twilight&ik-syntax-jupyter=ik-jp-twilight&ik-bg-choice=ik-yes&ik-bg-options=ik-Tiled&ik-bg-attachment=ik-Scroll&ik-tab-size=ik-4&ik-code-wrap=ik-nowrap&ik-base-color=%234183C4&ik-font-choice=Menlo&ik-bg-url=ik-default` for customizable styles.

In your link it's a raw CSS string that needs to be converted with our `moz-parser.js` (if you go this route, load it dynamically with `loadScript`). However you'll need to render the customizable variables e.g. `/*[[base-color]]*/` in [github dark](https://userstyles.org/styles/37035/github-dark). And we'll need to allow users to customize those variables (for example the way we do with usercss in the style manager) or at least tell them to do it on the style page.

tophf commented:
Also, install and configure latest eslint so that it uses our `.eslintrc` config - there are several minor issues in searchResults.js (hint: use `/* global name1, name2 */` comment as the other files do). It should also be named as `search-results.js` (same for the .css file). We have just one file with camelCase name inherited from Stylish which we all keep forgetting to rename even though the contents is 99.9% different.

tophf commented on popup/searchResults.js:
```diff
@@ -0,0 +1,251 @@
+   * @param {string} queryParams Query parameters to send in search request (e.g. "key=value&name=that)".
+   * @return {Object} Response object from userstyles.org
+   */
+  function fetch(path, queryParams) {
```
We already have `download()` and `tryJSONparse()`

derv82 replied:
Thanks for telling me about ESLint. It found a number of warnings/errors that I have since fixed. The ESLint fixes will show up in my next PR update.

---------

I think I can `download()`  the `/styles/chrome/*.json` file. And I'll use `tryJSONparse()` when possible.

However, `download()` [forces the method to `POST` if the URL has a `search` field](https://github.com/openstyles/stylus/blob/master/js/messaging.js#L387). The UserStyles search API (`/api/v1/styles/search?search=...&page=...`) requires URL Search Parameters, and the request has to be a `GET`.

Performing a `POST` on the search API returns a `405` error:

```bash
$ curl -X POST "https://userstyles.org/api/v1/styles/search?search=userstyles.org&page=1&per_page=2&category=userstyles.org" -vs 2>&1 | grep "< HTTP"
< HTTP/1.1 405 Method Not Allowed
```

I see [`content/install-user-css.js` has it's own `fetchText()` method](https://github.com/openstyles/stylus/blob/master/content/install-user-css.js#L7-L16). But it's encapsulated within `createSourceLoader`.

Should I:

1. Keep `UserStylesAPI.fetch()` in `search-results.js` since it can't use `download()`? Or,
2. Extract `fetchText()` into some utility class, to be used by both `install-user-style.js` and `search-results.js`?

I'm leaning towards Option 1 since the userstyles.org API may require specific headers in the future (e.g. `Content-Type`, `Accept`, `POST` instead of `GET`), and `fetchText()` doesn't currently support customizing headers.

tophf replied:
leave it as is, we'll deal with this later

Patelraj007 replied:
Bhai WiFi password hack show video give me

derv82 commented:
Thanks for the feedback.

I got `saveStyleSafe()` working, it marks installed styles as *Installed*. The style appears at the top of the `popup.html` immediately, which is cool.

There's still quite a few `TODO`s that I can tackle in time (or you can take over if you'd like):

1. Show weekly/total install counts for each search result
2. Detect installed styles from search results, and mark as such (or remove from Search Results entirely).
   * Any hints on identifying installed styles based on UserStyles `id`, `url`, or `name`?
3. Find a way to allow user customization. Seems like the options are:
   * Insert customization options into the SearchResult, above `Install` button
   * or display a message if the style is customizable, and hint at visiting the style page to customize.

tophf commented:
Nah, take your time. There's definitely no rush.

derv82 commented:
Getting close all of the features that what I wanted.

### DONE

* Installs style and updates the list of installed styles immediately.
* Improved search results.
  * Searches by-category (better than userstyles.org's search!)
  * Only shows results that apply to the current website.
* Only loads more search results when needed.
  * Although it's a bit slow: 500ms between each result appearing.
  * Can be configured via `DELAY_AFTER_FETCHING_STYLES`

### TODO

* Show install counts & rating for each result.
* Detect style customization & handle it in some way.
* Support hotkeys when style is installed via search results:
  * Right now, the "uninstall" trash can icon is not clickable until the popup is closed & reopened.
* Stop spamming `console.log` (currently left in for development purposes).

derv82 commented:
As I said in #243, I'm pretty happy with the result:

![stylus-searchresults-2](https://user-images.githubusercontent.com/993464/33523312-52dd376a-d7b8-11e7-8540-a9026c0928b4.gif)

Current code passes `eslint --fix-dry-run .` with 2 warnings from unrelated files (`vendor-overwrites\colorpicker\colorview.js` & `js\messaging.js`).

I think the PR is in a "good" state and ready to be reviewed.

Please note that the new `hotkeys.js` features broke the ability to uninstall a style (via the trashcan icon) after clicking `Install` on a search result... I looked into how hotkeys are loaded and it *looks like* they're loaded when popup.js finishes loading... but I didn't see any easy way to re-apply hotkeys (or whatever it does) when a style is installed *while the popup is still open*. If you have any ideas for a workaround, let me know.

tophf commented:
Don't emit any outgoing requests without an explicit user action such as clicking on the search link. This is what Stylish does. An acceptable approach would be to make two parallel requests on click and amend the results in that improbable case the category extracted via simple regexp was wrong.

tophf commented:
Also, there should be a way to open the search page just like before. Maybe rename the action to "Find styles" and add a checkbox to the right: "inline".

tophf commented on _locales/en/messages.json:
```diff
@@ -349,6 +349,38 @@
+    "description": "Text for label that shows the number of times a search result was installed"
+  },
+  "searchResultInstall": {
+    "message": "Install",
```
We already have `installButton` message

tophf commented on _locales/en/messages.json:
```diff
@@ -349,6 +349,38 @@
+    "description": "Text for a button that installs a search result"
+  },
+  "searchResultInstallTooltip": {
+    "message": "Install this search result",
```
No need for the tooltip.

tophf commented on _locales/en/messages.json:
```diff
@@ -349,6 +349,38 @@
+    "description": "Text for a button that customizes a search result before installing"
+  },
+  "searchResultCustomizeTooltip": {
+    "message": "Customize this style on userstyles.org",
```
Worth mentioning that it opens a page: `Open the style page on userstyles.org to customize it` or something

tophf commented on popup.html:
```diff
@@ -82,12 +83,39 @@
+      <a class="searchResult-title" target="_blank"></a>
+      <div class="searchResult-description"></div>
+      <div class="searchResult-author">
+        By:
```
1. untranslated label
2. should be inside an element

tophf commented on popup.html:
```diff
@@ -118,11 +146,33 @@
+        <span id="searchResults-terms">...</span>
+      </h3>
+      <div id="searchResultsNav">
+        <button id="searchResultsNav-prev" title="Previous page" disabled>Prev</button>
```
untranslated titles

tophf commented on popup/popup.css:
```diff
@@ -1,5 +1,7 @@
 html, body {
   height: min-content;
+  overflow-y: auto;
```
these should be applied only when the search container is shown e.g. it would add a class on `body` so a CSS rule for `body.search-results-shown` would be used

tophf commented on popup/popup.js:
```diff
@@ -112,8 +112,8 @@ function initPopup(url) {
   const u = tryCatch(() => new URL(url));
-  $('#find-styles-link').onclick = handleEvent.openURLandHide;
-  $('#find-styles-link').href +=
+  $('#open-search-link').onclick = handleEvent.openURLandHide;
```
the original element id shouldn't be changed; the behavior would be toggled by a checkbox to the right of this link; the second link displayed in the container may have a different id e.g. `find-more-styles-link`

tophf commented on popup/search-results.css:
```diff
@@ -0,0 +1,129 @@
+#find-styles.hidden,
+#open-search.hidden,
+.searchResult-customize.hidden {
+  display: none;
```
simply add one rule for `.hidden` just like on our other pages:

```css
.hidden {
  display: none !important;
}
```

tophf commented on popup/search-results.css:
```diff
@@ -0,0 +1,129 @@
+
+.searchResult > * {
+  overflow-x: hidden;
+  white-space: nowrap;
```
Hiding the overflow makes it hard to quickly rad the titles, let the title flow and add `text-overflow: ellipsis` on the description

tophf commented on popup/search-results.css:
```diff
@@ -0,0 +1,129 @@
+
+.searchResult-description {
+  overflow: hidden;
+  font-style: italic;
```
We don't use italic anywhere so it looks weird here

tophf commented on popup/search-results.css:
```diff
@@ -0,0 +1,129 @@
+}
+
+#searchResults, #searchResults-list {
+  max-width: 200px;
```
I think the width shouldn't be constrained as the white area on the right looks weird with the title and description cut off, especially if the popup width was customized to a bigger value. Instead let the description/installation block jump to the right when there's space e.g. if the popup width is bigger than 360px.

tophf commented on popup/search-results.js:
```diff
@@ -0,0 +1,547 @@
+    return new Promise(resolve => {
+      const request = new XMLHttpRequest();
+      const browseURL = 'https://userstyles.org/styles/browse/all/' + encodeURIComponent(url);
+      request.open('HEAD', browseURL, true);
```
you can use `download()` - it accepts custom request parameters now; the base URL should be a named constant at the top of `searchUserstyles` closure

tophf commented on popup/search-results.js:
```diff
@@ -0,0 +1,547 @@
+ * Library for interacting with userstyles.org
+ * @returns {Object} Exposed methods representing the search results on userstyles.org
+ */
+function SearchUserstyles() {
```
It's not a class factory so it should start with a lowercase letter

tophf commented on popup/search-results.js:
```diff
@@ -0,0 +1,547 @@
+   */
+  function fetchStyleJson(userstylesId) {
+    return new Promise((resolve, reject) => {
+      const jsonUrl = 'https://userstyles.org/styles/chrome/' + userstylesId + '.json';
```
the base URL should be a named constant at the top of `searchUserstyles` closure

tophf commented on popup/search-results.js:
```diff
@@ -0,0 +1,547 @@
+        'Accept': '*/*'
+      };
+
+      const styleUrl = new URL('https://userstyles.org/api/v1/styles/' + userstylesId);
```
the base URL should be a named constant at the top of `searchUserstyles` closure

tophf commented on popup/search-results.js:
```diff
@@ -0,0 +1,547 @@
+      };
+
+      const styleUrl = new URL('https://userstyles.org/api/v1/styles/' + userstylesId);
+      const xhr = new XMLHttpRequest();
```
It should be possible to use `download()` now

tophf commented on popup/search-results.js:
```diff
@@ -0,0 +1,547 @@
+        resolve({'data':[]});
+      }
+
+      const TIMEOUT = 10000;
```
move TIMEOUT to the top of `searchUserstyles` closure

tophf commented on popup/search-results.js:
```diff
@@ -0,0 +1,547 @@
+      queryParams += '&page=' + currentPage;
+      queryParams += '&country=NA';
+      searchUrl.search = '?' + queryParams;
+      const xhr = new XMLHttpRequest();
```
It should be possible to use `download()` now

tophf commented on popup/search-results.js:
```diff
@@ -0,0 +1,547 @@
+      };
+
+      const searchUrl = new URL('https://userstyles.org/api/v1/styles/subcategory');
+      let queryParams = 'search=' + encodeURIComponent(category);
```
Just use a single multiline statement.

tophf commented on popup/search-results.js:
```diff
@@ -0,0 +1,547 @@
+  }
+})();
+
+onDOMready().then(() => {
```
Move it to the top of the file so it's obvious what happens on including the js right away

tophf commented on popup/search-results.js:
```diff
@@ -0,0 +1,547 @@
+ * Represents the search results within the Stylus popup.
+ * @returns {Object} Includes load(), next(), and prev() methods to alter the search results.
+ */
+const SearchResults = (() => {
```
It's not reusable outside of this feature file so it should be inside one closure above

tophf commented on popup/search-results.js:
```diff
@@ -0,0 +1,547 @@
+   * @returns {Number} The *Next* page to fetch for styles. Auto-incremented after each search.
+   */
+  function getCurrentPage() {
+    return currentPage;
```
While it looks mature and all, but we don't need these getters outside of the closure. We can add them when it's actually needed, if ever.

tophf commented on popup/search-results.js:
```diff
@@ -0,0 +1,547 @@
+  return {load, next, prev};
+
+  function render() {
+    $('#searchResults-list').innerHTML = ''; // Clear search results
```
use `textContent = ''` - AMO dislikes any mentions to innerHTML so we at least avoid it in our own code

tophf commented on popup/search-results.js:
```diff
@@ -0,0 +1,547 @@
+    if (currentDisplayedPage <= 1 || loading) {
+      $('#searchResultsNav-prev').setAttribute('disabled', 'disabled');
+    } else {
+      $('#searchResultsNav-prev').removeAttribute('disabled');
```
Use `.disabled = currentDisplayedPage <= 1 || loading`, this is a standard DOM property.

tophf commented on popup/search-results.js:
```diff
@@ -0,0 +1,547 @@
+    }
+    const totalPageCount = Math.ceil(Math.max(1, totalResultsCount / DISPLAYED_RESULTS_PER_PAGE));
+    if (currentDisplayedPage >= totalPageCount || loading) {
+      $('#searchResultsNav-next').setAttribute('disabled', 'disabled');
```
same

tophf commented on popup/search-results.js:
```diff
@@ -0,0 +1,547 @@
+
+    const navNode = $('#searchResultsNav');
+    if (loading && !navNode.classList.contains('loading')) {
+      navNode.classList.add('loading');
```
use `.toggle` with a second parameter instead of if/else:
`navNode.classList.toggle('loading', loading);`

tophf commented on popup/search-results.js:
```diff
@@ -0,0 +1,547 @@
+      message = 'Error loading search results: ' + reason;
+    }
+    $('#searchResults').classList.add('hidden');
+    $('#searchResults-error').innerHTML = message;
```
Use `.textContent`

tophf commented on popup/search-results.js:
```diff
@@ -0,0 +1,547 @@
+    }
+
+    $('#find-styles').classList.add('hidden');
+    $('#open-search').classList.remove('hidden');
```
While the steps look good, but the states should be grouped; move one of them so that `add`'s are together. Might even add an empty line between.

tophf commented on popup/search-results.js:
```diff
@@ -0,0 +1,547 @@
+            let isMatch = installedStyle.name === userstyleSearchResult.name;
+            // Compare if search result ID (userstyles ID) is mentioned in the installed updateUrl.
+            if (installedStyle.updateUrl) {
+              isMatch &= installedStyle.updateUrl.indexOf('/' + userstyleSearchResult.id + '.json') >= 0;
```
use `.includes()` instead of `.indexOf() >= 0`

tophf commented on popup/search-results.js:
```diff
@@ -0,0 +1,547 @@
+    if (screenshotUrl === null) {
+      screenshotUrl = BLANK_PIXEL_DATA;
+    } else if (RegExp(/^[0-9]*_after.(jpe?g|png|gif)$/i).test(screenshotUrl)) {
+      screenshotUrl = 'https://userstyles.org/style_screenshot_thumbnails/' + screenshotUrl;
```
Make the base URL a named constant and move it to the top of the main closure

tophf commented on popup/search-results.js:
```diff
@@ -0,0 +1,547 @@
+    Object.assign(title, {
+      textContent: searchResultName + ' (by ' + userstyleSearchResult.user.name + ')',
+      title: searchResultName + ' by: ' + userstyleSearchResult.user.name,
+      href: 'https://userstyles.org' + userstyleSearchResult.url,
```
Make the base URL a named constant and move it to the top of the main closure

tophf commented on popup/search-results.js:
```diff
@@ -0,0 +1,547 @@
+    Object.assign(authorLink, {
+      textContent: userstyleSearchResult.user.name,
+      title: userstyleSearchResult.user.name,
+      href: 'https://userstyles.org/users/' + userstyleSearchResult.user.id,
```
Make the base URL a named constant and move it to the top of the main closure

tophf commented on popup/search-results.js:
```diff
@@ -0,0 +1,547 @@
+      installButton.classList.add('customize');
+      const customizeButton = $('.searchResult-customize', entry);
+      customizeButton.classList.remove('hidden');
+      customizeButton.href = 'https://userstyles.org' + userstyleSearchResult.url;
```
Use a named constant for the base URL

tophf commented on popup/search-results.js:
```diff
@@ -0,0 +1,547 @@
+      entry.classList.add('loading');
+
+      const styleId = userstyleSearchResult.id;
+      const url = 'https://userstyles.org/styles/chrome/' + styleId + '.json';
```
Make the base URL a named constant and move it to the top of the main closure

tophf commented:
Overall, despite the multiple disapproving comments I made above, the feature is great and the code is neat. Don't be disheartened by my nitpicking.

tophf commented:
>Right now, the "uninstall" trash can icon is not clickable until the popup is closed & reopened.

Works4me.

derv82 commented:
Sweet, this is the kind of feedback I was looking for.

I'll try to address all of the comments & let you know when it's ready for another review (just ignore any commits I push for now).

If I have any questions, I'll respond above inline or ping the Discord.

derv82 commented:
> and add a checkbox to the right: "inline".

I liked this idea. But adding a checkbox to the right of `find styles for this site` has proven difficult due to the SVG checkbox system used on popup.html.

I tried creating this:

* `Find styles for this site` `[ ] inline`
  * `inline` is the checkbox label

But apparently the SVG checker position is **hard-coded to the pixel** in popup.css, which explains why all checkboxes in popup.html appear on the far-left. Checking `inline` box causes the "checker" icon to appear floating elsewhere in the popup (not over the `[ ] inline` checkbox area.

I *could* specify checker position (e.g. `top:40px; left:90px`) but the position of the checkbox is subject to change based on the length of the i18n-text (varies by language).

I'm debating how to resolve this. I can do:

* `[ ] Inline` `Find styles for this site`
  * `Inline` is the checkbox label
  * `Find styles for this site` is the anchor link

But that isn't very intuitive.

Or I could just duplicate the anchor links:

* `Find styles for this site (Inline)`
* `Find styles for this site (New Tab)`

Which is ugly.

Or I could specify the `inline` checker position when checkbox/label is clicked (`onmousedown` event) by reading the position of the checkbox. Which is pretty hacky but would be consistent.

Any ideas?

tophf commented:
1. Shorten the label to "Find styles"
2. Wrap the checkbox in a relatively positioned label e.g.
  `<label>[checkbox stuff]<span></span></label>`
  The translation attribute `i18n-text` should be on the span. The elements should have some classes/ids.
3. Add any necessary CSS overrides

derv82 commented:
Still working on the other feedback (not ready for full review) but you can take a look at how I got the `Inline` checkbox positioning... Hopefully this isn't a cardinal sin:

```css
#find-styles-inline-group input[type="checkbox"],
#find-styles-inline-group .svg-icon.checked {
  position: relative;
}

#find-styles-inline-group .svg-icon.checked {
  margin-left: -13px;
  margin-right: 2px;
}
```

narcolepticinsomniac commented:
`position: relative;` belongs on `label`. `label` should be the parent to the checkbox. No extra positioning CSS should be necessary.

narcolepticinsomniac commented:
If the global CSS doesn't add it, put `padding-left: 16px;` on the `label`.

tophf commented:
Oops I forgot.  I think the short "Find styles" in messages.json should be have a new message id, while the old one will be used in 
`i18n-title` for the element. BTW we can tweak CSS later if it's not your thing :-)

derv82 commented:
@narcolepticinsomniac Thanks for the tip; that works well.

> BTW we can tweak CSS later if it's not your thing :-)

That would be great; positioning & UI flow are not my strong points. I could learn some things in the these areas.

--------------

# From PR Feedback

### search-results.js:
* Ctrl-click/Right-click on `Find styles` link will open new tab to USO search page, even if `inline` is checked.
* Renamed `SearchUserstyles` => `searchUserstyles`
* Removed unused "getters"
* Always use `download()` (no more `XMLHttpRequest` or `TIMEOUT`)
* Use a `BASE_URL` constant
* `searchURL` built using multiline statement
* Moved `onDOMready()` to top of file
* Renamed `SearchResults` to `searchResultsController`; lives inside IIFE under `onDOMready()`.
* Use `.textContent = ...` instead of `.innerHTML = ...`
* Consolidated `.disabled` if-else to single-line boolean expression.
* `loading` var moved to `setLoading()` function; deals with current loading state.
* `.includes(..)` instead of `.indexof(...) >= 0`

### popup.html
* Entire search result is clickable -- opens new tab for the style
  * Interactive elements ("Install", "Customize", description expander) stop click propagation.
* "Next" and "Prev" buttons are now labelled with arrows (HTML entities)
  * `i18n-titles` for tooltips on next/prev buttons and current/total labels.
* Shows `.lds-spinner` on top of search result pane while loading via network.

### popup.js
* Moved `#find-styles-link` onclick logic (inline/not-inline) to search-results.js

### popup.css
* Only apply 'overflow' to `body.search-results-shown`
* Consolidated to use `.hidden`
* Search results no longer constrain to a max-width; grow to size of popup window width.
* No italics
* Search result title no longer overflows
* Set right-padding when search results are shown to account for scrollbar
  * (This is why I couldn't uninstall styles while viewing results)
  * There's probably a better way to do this...

### i18n
* Shortened 'Find styles for site' to 'Find styles'
* All labels should have `i18n-text`
* No titles without `i18n-title`

---------

# Outstanding Problems

### "Description expander" is terrible
* Uses the `(i)` info icon.
* Description text is `width:90%` to allow room for `(i)` icon.
* Clicking `(i)` removes description overflow
  * Should probably hover a new element containing the full description, or show a modal with full descrition.

### `.lds-spinner` duplicated in `search-results.css`
* I was worried about including `install-usercss.css` into the `popup.html` window...
* Spinner does not completely cover the search result area.
  * Should have separately loaders for each not-yet-processed result.

tophf commented:
Okay, now that it works it's time to discuss the way it'll be used.

I doubt users want to open the USO page on clicking the thumbnail. I think the most obvious and expected action is to install the style. It's like instant previewing. The user can uninstall it easily by clicking the trashcan icon. The only minor nuisance is the need to scroll the popup, but you can circumvent it by displaying another trashcan icon over the thumbnails of styles installed in this popup session.

This is what I think would look and feel better:

* Name before thumbnail
* Install button shown on hovering the thumbnail
* Rating and install count shown on hover within the thumbnail area on a semitransparent background e.g. 50% white
* Description shown on hover so it overlays/replaces the thumbnail
* No highlighting of the card's background on hover as it won't be needed anymore.
* Remove "Results for xxxxxx", just list the results.
* Insert the results at the end of the popup in order to preserve the layout structure.

So basically it'd be the title and the thumbnail at first glance. Simple and clear. We can add a checkbox [later] to always show full descriptions or we can leave it userstyle authors.

Some trivial issues: 

* use dashes in the bunch of `searchResult` classes instead of camelCase
* when the spinner appears for the first time it overlays the main buttons.

As for

derv82 commented:
Here's my summary of your suggestions with my interpretation/plan to implement:

### Search-Result onclick
> I think the most obvious and expected action is to install the style. 
> [...]
> The only minor nuisance is the need to scroll the popup, but you can circumvent it by displaying another trashcan icon over the thumbnails of styles installed in this popup session.

:+1: Interpreted as:

* Clicking search result installs it.
* Installing doesn't remove result from list:
  * Instead, place trashcan icon over thumbnail. Uninstall when clicked.

----------

### Hover behavior
**"Not Hovering"**
> So basically it'd be the title and the thumbnail at first glance.

:+1: Interpreted as:
* When not hovering, only show Search Result's name & thumbnail:

```
Dark Github
+-----------------+
|                 |
|   Screen-shot   |
|                 |
+-----------------+
```

------------

**"On Hover"**

> * No highlighting of the card's background on hover as it won't be needed anymore.
> * Install button shown on hovering the thumbnail
> * Rating and install count shown on hover within the thumbnail area
> * Description shown on hover so it overlays/replaces the thumbnail

:hand: I need some clarification:
* Should description stay 1-line with the `(i)` expando?
  * Showing entire description probably won't fit within the thumbnail area of many styles
  * That might cause a nasty overflow bleeding into other styles (I think I have to `position:absolute` the overlay)...
  * Are we OK with truncating the description inside the thumbnail? If so, do we need the `(i)` expando?
* Thumbnail area might get too crowded with install button, ratings/installs, and description.
  * I also wanted to add a "Last Updated" label (number of days since style was updated)

Interpreted as:

* When hovering a thumbnail, show an overlay on the thumbnail (`bg:white; opacity:0.5`)
* Inside overlay:
  * Show "Install" button
  * Show Rating & Install count
  * Show description with overflow-y:hidden

```
Dark Github
+---------------------+
|      [Install]      |
| Rating:x Installs:y |
|    (Description)    |
+---------------------+
```

----------

> * Insert the results at the end of the popup in order to preserve the layout structure.

:+1:
* But should `Find styles` *also* be at the end of the popup (but above the search results)?
  * I'm worried the action -> result distance might be too big, especially if more options are added to `popup.html` (between `Find styles` & end of popup) in the future.

----------

> * Remove "Results for xxxxxx", just list the results.
> * use dashes in the bunch of searchResult classes instead of camelCase

:+1: Got it.

----------

> * when the spinner appears for the first time it overlays the main buttons.

😨 I'll try to fix it.

-----------

Hopefully in the next few iterations, the only remaining work will be styling & positioning that can be fixed by CSS. I'm hoping someone else (CSS guru) can take over at that point, otherwise we'll burn time and precious Github comments with:

1. Suggestion
2. My attempt to implement using 1990's-era CSS
3. Suggestion to implement in a better way
4. I copy/paste.

I foresee a wave of new Stylus userstyles designed to fix the mess I've made 😐 .

----------

> As for
> `<EOF>`

😰

tophf commented:
User-friendly install-on-click-on-thumbnail is implemented in Stylish last time I saw it - a text label appears in the center of the thumbnail if I'm not mistaken, and the entire thumbnail area is clickable.

>I'm worried the action -> result distance might be too big, especially if more options are added to popup.html

It's the other way around: we'll reduce that area eventually. Also not a problem either way - just scroll the popup to the top of the results container.

>Thumbnail area might get too crowded with install button, ratings/installs, and description

Hm. We need to figure something out. Currently there's too much info that's hard to read in such a small area. Maybe we should expand the popup width temporarily to 450-500px to fit the descriptions and info on the right. I don't know yet.

>I foresee a wave of new Stylus userstyles designed to fix the mess I've made

We'll certainly do our best to make sure it's not a mess before we release it.

----------

Why `padding-right: 20px;` in `body.search-results-shown` rule? It just breaks the layout. If you want to widen the popup, change its document.body.style.width.

derv82 commented:
> Why `padding-right: 20px;` in `body.search-results-shown` rule? It just breaks the layout.

My browser's/OS's scrollbar covered part of the page when popup.html height is > some value.

Apparently I only needed `overflow-y: scroll`. No need to specify width. Whoops!

`overflow-y: auto` (No `padding-right`) | `overflow-y: scroll`
---------------------|---------------------
![stylus-no-body-padding-right](https://user-images.githubusercontent.com/993464/33707875-b5d02368-daed-11e7-94b6-7a6f1e386247.gif) | ![stylus-body-overflow-y-scroll](https://user-images.githubusercontent.com/993464/33707917-e244b850-daed-11e7-9972-2dffb3eccf61.gif)

---------

Here's how the overlay looks (briefly teased in the above GIF):

![stylus-searchresult-overlay](https://user-images.githubusercontent.com/993464/33708185-9437cd18-daee-11e7-852b-58522b5aa806.PNG)

-------

I addressed all of the points, except the "Uninstall" option -- still removes the search result after installing. I'll update PR once "Uninstall"/trashcan functionality is added.

tophf commented:
Wow, I think now it's almost awesome. The only weird thing left is that idea with covering the entire thumbnail doesn't seem to be as good as I thought. How about you cover just the bottom 50%? All that info can be compressed I think. The install button doesn't need to cover the entire width, let it fit the text, let it sit somewhere in the middle maybe because it's the focal point.

derv82 commented:
Latest changes in GIF form:

![stylus-searchresult-halfoverlay](https://user-images.githubusercontent.com/993464/33751610-e930527a-db90-11e7-9f81-cbb52597bd6f.gif)

tophf commented:
Yeah, definitely better. I would lower it even more so that the install button "floats in the space" by itself, what do you think?

tophf commented:
I also think there's no need to reduce the popup's height when navigating pages because it looks jumpy. Just fade out the currently displayed results to `opacity: 0` and maybe display the spinner in the middle with transition from `opacity: 0` to `opacity: 1`.

tophf commented:
Three styles seems insufficient now that I've been using the feature for a while. I think 10 would be better. You mentioned you're using the full style code for the added reliability of checking whether the styles apply, but it's probably needed only in a few edge cases where the broken USO engine produces irrelevant results for google/whatever. If 99% of sites don't need that then add the exclusions for the known buggy matches for which you will fetch the full code whereas the majority of sites will only need one small json with the search results.

derv82 commented:
I'll try moving "Install" button to the area above the overlay, and moving the overlay down more.

I'll bump the search results to 10 and insert placeholder results (empty except for an `lds-spinner`) for not-yet-processed results to avoid height-changes (jumpiness).

------

> If 99% of sites don't need that then add the exclusions for the known buggy matches for which you will fetch the full code whereas the majority of sites will only need one small json with the search results.

Refactoring the search-result-processing code is gonna be a pain so I want to get this sorted before making changes.

Right now, before each search result is displayed, we make two asynchronous calls:
  * `fetchStyleJson()`: As you said, this purely for detecting if style is applicable to the URL. We *do* reuse the JSON when installing (which is why installation is so fast)
  * `fetchStyle()`: For detecting "style_settings". So we wouldn't know if style is customizable (lose the "Customize" button, which was kind of a nice feature).

Removing **both** calls at search-time:
* Pros:
  * Super-fast search results.
  * We're limited to 15 results at-a-time (limitation of category-based searching), but we can "preload" the next page of results (I had done this in an earlier iteration).
  * The Next button would be briefly disabled while "preloading", but otherwise it's practically invisible to the user.
* Cons:
  * Potentially-non-applicable search results (likely not a problem as you said, now that we have a working category-based search).
  * Delay when installing styles (have to wait for `fetchStyleJson`).
  * No more "Customize" button (slight feature-loss).

If we keep **either** of the calls, the results will continue to be slow-loading as they are now. I want to keep these requests synchronous to avoid DDoSing USO (or getting IP banned / blacklisted).
  * If we went asynchronous, that'd be `(1 search) + (15 fetchStyleJson) + (15 fetchStyle) = 31` network requests just by clicking `Find styles` (`62` if we "preload").

Once I know what features are desired (fast results vs. accurate results + Customize + fast installation), I can refactor the processing logic, search-result-preloading, asynchronicity, etc.

tophf commented:
Well, my point is that always fetching the entire style code is a waste of user's and USO bandwidth in most of the cases. 

* It's also quite unusual, although such a bug-ridden site as popular as USO is also unusual. 
* I suspect the users would only install 0 to 2 styles per each invocation of the search function. At least after they've done playing with the new feature. 
* On-demand installing is quite fast, and the delay isn't something unexpected especially if you fade in the spinner after 100-200ms of waiting.
* The customize button may be shown after the style is installed and it would be good UX. Who wants to customize a style without trying its default look, anyway?
* As for accuracy, I don't know. We could add an option at the bottom of the search container e.g. `[x] Double-check the results`, but is it really needed? Provided the straightforward search works in 99% of cases. For the bad sites you can continue with the current approach I guess.

>Refactoring the search-result-processing code is gonna be a pain

We can merge as is and I'll deal with the issues myself.

derv82 commented:
Commit says it all.

I couldn't get height to stay put while loading search results (might be 0 or 10). A single spinner is shown while loading search results. I tried `min-height` on the search-results-list, but then < 10 results had a huge gap at the bottom.

However, height remains steady once the number of results are known (i.e. while fetching JSONs of each style) via the use of "loading" placeholder elements.

tophf commented:
With the exception of some CSS glitches this is ready to be tested extensively by more people, @narcolepticinsomniac, @Mottie, @MechaLynx. We can fix those glitches after merging. The only big problem left, IMO, is that it's still fetching the full style code for all displayed styles prior to clicking the install button, which it shouldn't.

Mottie commented:
* There is a huge empty space under the manage button block.
* The opacity of the info block needs to be increased. It can be difficult to read with some screenshot images. And make the (i) icon easier to see.

    ![](https://user-images.githubusercontent.com/136959/33780290-e2b268c2-dc15-11e7-81b5-7fd884d34a62.png)

* Clicking on the (i) icon doesn't show the full description, it might be better to add *all* of the description into the title of the (i) instead of making it clickable.
* When clicking on the next page button, hovering over any style shows flickering while loading. It looks like a simple css hover is being used, so I'm not sure about an easy way to solve this issue.

    ![](https://user-images.githubusercontent.com/136959/33780585-e8acb600-dc16-11e7-870a-b81d2ec818e9.gif)

tophf commented:
We'll fix all CSS issues later (e.g. add a transition delay on hover or adjusting margins and such should be trivial, at least for us). What I want is more people to use this feature for real while testing.

Mottie commented:
And maybe the "Find Styles" would look better as a button. And add an external link icon to indicate finding styles on USO.

Mottie commented:
* It would be nice to keep a few pager buttons at the bottom, or at least a button to go back to the first page without needing to close and reopen the popup. I'm guessing the API doesn't provide a total number of styles?
* I'm not sure why I'm seeing only 2 pages with GitHub styles, then USO shows a lot more. I see you said a limit of 10 styles is applied, but when I look at Google's home page, it goes way beyond 10.

Mottie commented:
* If I delete a style using the entries at the top of the popup, it doesn't update the delete button shown over the thumbnail.

derv82 commented:
> The only big problem left, IMO, is that it's still fetching the full style code for all displayed styles prior to clicking the install button, which it shouldn't.

Moved `fetchStyleJson()` call to the `install()` function.
* JSON is only fetched when install is clicked
* While installing, the "Install" button is disabled & there's a spinner.
* Note: No longer checks if style is applicable for current URL.

TBH I think most of the CSS problems (e.g. flickering) are due to the use of `render()` to modify the search results. `render()` wipes out the contents of `#search-results-list` and rebuilds the whole list on every call. So the `opacity` transition gets triggered on every `render()` call.

I'll wait for @tophf feedback on what to do next.

tophf commented:
I think I can merge it and deal with the issues myself if you don't mind.

derv82 commented:
Please do!

tophf commented:
4b207f3dd4b18de50a48ea5daec1c25bd1dd8272 fixes some stuff that irked me personally.

narcolepticinsomniac commented:
Nice work guys! 

Overlays can be glitchy when loading "next page" of results, because after clicking the button, you may already hovering a style as it's loading. Possible solution would be an overlay for the whole list while it's loading, until it's fully loaded: Glitch example: [overlay.zip](https://github.com/openstyles/stylus/files/1545262/overlay.zip)

Overlay positioning LGTM. Have you experimented with light text on dark overlay backgrounds? Always a more polished look IMO.

tophf commented:
@narcolepticinsomniac you're looking at the old code. The new one is in the master branch.

narcolepticinsomniac commented:
My bad. The compliment stands though. Thought Mottie used a real recent master. I'll check it out.

Mottie commented:
I tried too! @tophf has been pushing commits like crazy today.

narcolepticinsomniac commented:
Cool, no more glitch. You did switch some of the overlay bg to light text on dark. Not sure why the titles aren't included in that. I would think a simple semitransparent black with light text under the entire bottom overlay would look nice. Maybe a little text shadow for legibility. The orange rating bg is kinda fugly. 

While better, we still have some extra space under `#disable-all-wrapper` and at the bottom, under `#popup-options` that look a little weird.

tophf commented:
The titles have white background because the card is white and lots of screenshots are white. I don't know the ratio of light/dark styles on USO but I wouldn't be surprised if the dark ones are a minority, numbers-wise. I'm having doubts even about the darkcyan overlay I've used.

The orange ratings are fugly because it's a low rating I guess. I've seen red too. Good ones are greenish.

tophf commented:
As for the wrapper, the PR broke some html by adding another .actions block, I'm gonna fix it now.

narcolepticinsomniac commented:
>I'm having doubts even about the darkcyan overlay I've used

If done well, light text on black with text shadow work on top of anything. You see it used everywhere in web design.

I thought it might be nice if "configure" URL added a hash to focus "advanced settings", since they've kinda buried them. Holy shit, have you looked at the US.o code? The level of stupidity is astounding. Ironic that a userstyle extension/website is now being run by people with no basic grasp of CSS. That's not how IDs work. At all.

tophf commented:
>I thought it might be nice if "configure" URL added a hash to focus "advanced settings", since they've kinda buried them

Yeah, me too! Gonna add yet another workaround to the site hook.

narcolepticinsomniac commented:
Titles have an overflow issue:

![overflow](https://user-images.githubusercontent.com/14100003/33801732-85d3ae04-dd32-11e7-912f-fb069a7bdc44.png)

tophf commented:
Yep, forgot to apply that hyphenation patch.

narcolepticinsomniac commented:
I get why we use it in the manager, but wouldn't ellipsis suffice for the popup?

narcolepticinsomniac commented:
NM, it's the long string with no spaces thing.

tophf commented:
That's ok for the installed styles, but from what I see there's just too many long titles to cut them.

narcolepticinsomniac commented:
Unlike installed styles, a hover tooltip would work well for ellipsis there, if we wanted to keep listings all the same size. Not the biggest deal if you really prefer it.

tophf commented:
I doubt that. I hate having to hover on basic info. Descriptions are ellipsized and it's fine as it's not a primary element. Cutting known title in installed styles is fine. We see them thousands of times. Cutting unknown titles in search results to have a simpler layout is bad.

narcolepticinsomniac commented:
That logic checks out. You're probably right. Overflow js then.

narcolepticinsomniac commented:
Overflow from inline "find styles" currently places scrollbars on the popup itself. With a dark style and "add borders" enabled, borders are inside the scrollbars, which isn't a good look:

![popup-overfow](https://user-images.githubusercontent.com/14100003/33863627-aabe3da8-deb6-11e7-8644-f874da068fa5.png)

Since there's a hard limit on popup height, we can fix it by adding a max-height:

```
html, body {
  height: min-content;
  max-height: 600px;
}
```

narcolepticinsomniac commented:
Hmm. There's an issue with `#hotkey-info` though.

narcolepticinsomniac commented:
Why is `#hotkey-info` not inside `#installed` and positioned according to it?

tophf commented:
Because it's not a style entry. Having it inside the container would be real PITA. Like real.

tophf commented:
We can make an additional wrapper though, which I was reluctant to do as it's a change of the basic layout.

narcolepticinsomniac commented:
Never paid much attention to its code. Why is it necessary at all? Couldn't numbers be actual elements with the same mouse change/click action on them?

tophf commented:
They could be, but having CSS deal with autonumbering when the styles are installed/deleted or change their order seemed an easier option.

narcolepticinsomniac commented:
Not trying to suggest complicating things. I'm sure you had your reasons. Any solution would do, I guess.

narcolepticinsomniac commented:
A class is added to the body, so we can reposition accordingly, I'd think.

tophf commented:
The problem seems harder to solve than I thought. The root cause is that the entire `body` can scroll with inline results displayed. So AFAICT it's not caused by `#hotkey-info` or any CSS in particular. I don't know how to solve this while keeping the ability to scroll the entire popup freely and have proper autoadjustment of popup height and the white borders where they should be.

tophf commented:
We could hide the native scrollbars and draw our own, but it seems a bit too much to tame the borders issue.

narcolepticinsomniac commented:
Well, it's not some emergency that needs to be fixed immediately, so we can take a little time to come up with the best solution.
